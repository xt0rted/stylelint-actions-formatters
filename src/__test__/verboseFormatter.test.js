/**
 * Based on https://github.com/stylelint/stylelint/blob/14.12.0/lib/formatters/__tests__/verboseFormatter.test.js
 */
'use strict';

const { stripIndent } = require('common-tags');

const prepareFormatterOutput = require('./prepareFormatterOutput');
const verboseFormatter = require('../verboseFormatter');

describe('verboseFormatter', () => {
  beforeAll(() => {
    jest.spyOn(process, 'cwd').mockReturnValue('/test-project');

    // This is needed when running on github actions
    delete process.env.GITHUB_WORKSPACE;
  });

  describe('using cwd', () => {
    runTestSuite();
  });

  describe('using GITHUB_WORKSPACE', () => {
    beforeAll(() => {
      process.env.GITHUB_WORKSPACE = '/test-project/path';
    });

    afterAll(() => {
      delete process.env.GITHUB_WORKSPACE;
    });

    runTestSuite();
  });

  function runTestSuite() {
    it('outputs no warnings', () => {
      const results = [
        {
          source: '/test-project/path/to/file.css',
          errored: false,
          warnings: [],
          deprecations: [],
          invalidOptionWarnings: [],
        },
      ];

      const output = prepareFormatterOutput(results, verboseFormatter);

      expect(output).toMatchSnapshot();
    });

    it("outputs one warnings (of severity 'error')", () => {
      const results = [
        {
          source: '/test-project/path/to/file.css',
          errored: true,
          warnings: [
            {
              line: 1,
              column: 2,
              rule: 'bar',
              severity: 'error',
              text: 'Unexpected foo',
            },
          ],
          deprecations: [],
          invalidOptionWarnings: [],
        },
      ];

      const output = prepareFormatterOutput(results, verboseFormatter);

      expect(output).toMatchSnapshot();
    });

    it('outputs 0 stdout column', () => {
      const stdoutColumn = process.stdout.columns;

      process.stdout.columns = 0;

      const results = [
        {
          source: '/test-project/path/to/file.css',
          errored: true,
          warnings: [
            {
              line: 1,
              column: 2,
              rule: 'bar',
              severity: 'error',
              text: 'Unexpected foo',
            },
          ],
          deprecations: [],
          invalidOptionWarnings: [],
        },
      ];

      const output = prepareFormatterOutput(results, verboseFormatter);

      expect(output).toMatchSnapshot();

      process.stdout.columns = stdoutColumn;
    });

    it('outputs less than 80 stdout column', () => {
      const stdoutColumn = process.stdout.columns;

      process.stdout.columns = 79;

      const results = [
        {
          source: '/test-project/path/to/file.css',
          errored: true,
          warnings: [
            {
              line: 1,
              column: 2,
              rule: 'bar',
              severity: 'error',
              text: 'Unexpected foo',
            },
          ],
          deprecations: [],
          invalidOptionWarnings: [],
        },
      ];

      const output = prepareFormatterOutput(results, verboseFormatter);

      expect(output).toMatchSnapshot();

      process.stdout.columns = stdoutColumn;
    });

    it("outputs two of the same warnings of 'error' and one of 'warning' across two files", () => {
      const results = [
        {
          source: '/test-project/path/to/file.css',
          errored: true,
          warnings: [
            {
              line: 1,
              column: 2,
              rule: 'bar',
              severity: 'error',
              text: 'Unexpected foo',
            },
            {
              line: 2,
              column: 3,
              rule: 'bar',
              severity: 'error',
              text: 'Unexpected foo',
            },
          ],
          deprecations: [],
          invalidOptionWarnings: [],
        },
        {
          source: '/test-project/path/to/file2.css',
          errored: true,
          warnings: [
            {
              line: 3,
              column: 1,
              rule: 'baz',
              severity: 'warning',
              text: 'Expected cat',
            },
          ],
          deprecations: [],
          invalidOptionWarnings: [],
        },
      ];

      const output = prepareFormatterOutput(results, verboseFormatter);

      expect(output).toMatchSnapshot();
    });

    it('outputs lineless syntax error', () => {
      const results = [
        {
          source: '/test-project/path/to/file.css',
          errored: false,
          warnings: [
            {
              rule: 'SyntaxError',
              severity: 'error',
              text: 'Unexpected foo',
            },
          ],
          deprecations: [],
          invalidOptionWarnings: [],
        },
      ];

      const output = prepareFormatterOutput(results, verboseFormatter);

      expect(output).toMatchSnapshot();
    });

    it('outputs one ignored file', () => {
      const results = [
        {
          source: 'file.css',
          warnings: [],
          deprecations: [],
          invalidOptionWarnings: [],
          ignored: true,
        },
      ];

      const output = prepareFormatterOutput(results, verboseFormatter);

      expect(output).toMatchSnapshot();
    });

    it('outputs input CSS', () => {
      const results = [
        {
          source: '<input css>',
          errored: true,
          warnings: [
            {
              line: 1,
              column: 2,
              rule: 'bar',
              severity: 'error',
              text: 'Unexpected foo',
            },
          ],
          deprecations: [],
          invalidOptionWarnings: [],
        },
      ];

      const output = prepareFormatterOutput(results, verboseFormatter);

      expect(output).toMatchSnapshot();
    });

    it('outputs plugin rule warnings', () => {
      const results = [
        {
          source: '/test-project/path/file.css',
          errored: true,
          warnings: [
            {
              line: 1,
              column: 2,
              rule: 'plugin/bar',
              severity: 'error',
              text: 'Unexpected foo',
            },
          ],
          deprecations: [],
          invalidOptionWarnings: [],
        },
      ];

      const output = prepareFormatterOutput(results, verboseFormatter);

      expect(output).toMatchSnapshot();
    });

    it('outputs rule warnings with metadata', () => {
      const results = [
        {
          source: '/test-project/path/file.css',
          errored: true,
          warnings: [
            {
              line: 1,
              column: 2,
              rule: 'no-foo',
              severity: 'error',
              text: 'Unexpected foo',
            },
          ],
          deprecations: [],
          invalidOptionWarnings: [],
        },
      ];
      const returnValue = {
        ruleMetadata: {
          'no-foo': { url: 'https://stylelint.io', fixable: true },
        },
      };

      const output = prepareFormatterOutput(results, verboseFormatter, returnValue);

      expect(output).toMatchSnapshot();
    });
  }
});
