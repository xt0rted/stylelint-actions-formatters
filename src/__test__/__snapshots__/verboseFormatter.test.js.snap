// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`verboseFormatter using GITHUB_WORKSPACE outputs 0 stdout column 1`] = `
"to/file.css
 1:2  ×  Unexpected foo  bar

1 problem (1 error, 0 warnings)

1 source checked
 /test-project/path/to/file.css

1 error found
 bar: 1"
`;

exports[`verboseFormatter using GITHUB_WORKSPACE outputs input CSS 1`] = `
"<input css>
 1:2  ×  Unexpected foo  bar

1 problem (1 error, 0 warnings)

1 source checked
 <input css>

1 error found
 bar: 1"
`;

exports[`verboseFormatter using GITHUB_WORKSPACE outputs less than 80 stdout column 1`] = `
"to/file.css
 1:2  ×  Unexpected foo  bar

1 problem (1 error, 0 warnings)

1 source checked
 /test-project/path/to/file.css

1 error found
 bar: 1"
`;

exports[`verboseFormatter using GITHUB_WORKSPACE outputs lineless syntax error 1`] = `
"to/file.css
      ×  Unexpected foo  SyntaxError

1 problem (1 error, 0 warnings)

1 source checked
 /test-project/path/to/file.css

1 error found
 SyntaxError: 1"
`;

exports[`verboseFormatter using GITHUB_WORKSPACE outputs no warnings 1`] = `
"1 source checked
 /test-project/path/to/file.css

0 problems found"
`;

exports[`verboseFormatter using GITHUB_WORKSPACE outputs one ignored file 1`] = `
"0 of 1 source checked
 file.css (ignored)

0 problems found"
`;

exports[`verboseFormatter using GITHUB_WORKSPACE outputs one warnings (of severity 'error') 1`] = `
"to/file.css
 1:2  ×  Unexpected foo  bar

1 problem (1 error, 0 warnings)

1 source checked
 /test-project/path/to/file.css

1 error found
 bar: 1"
`;

exports[`verboseFormatter using GITHUB_WORKSPACE outputs plugin rule warnings 1`] = `
"file.css
 1:2  ×  Unexpected foo  plugin/bar

1 problem (1 error, 0 warnings)

1 source checked
 /test-project/path/file.css

1 error found
 plugin/bar: 1"
`;

exports[`verboseFormatter using GITHUB_WORKSPACE outputs rule warnings with metadata 1`] = `
"file.css
 1:2  ×  Unexpected foo  no-foo

1 problem (1 error, 0 warnings)

1 source checked
 /test-project/path/file.css

1 error found
 no-foo: 1 (maybe fixable)

You may fix some problems with the \\"--fix\\" option."
`;

exports[`verboseFormatter using GITHUB_WORKSPACE outputs two of the same warnings of 'error' and one of 'warning' across two files 1`] = `
"to/file.css
 1:2  ×  Unexpected foo  bar
 2:3  ×  Unexpected foo  bar

to/file2.css
 3:1  ‼  Expected cat  baz

3 problems (2 errors, 1 warning)

2 sources checked
 /test-project/path/to/file.css
 /test-project/path/to/file2.css

2 errors found
 bar: 2

1 warning found
 baz: 1"
`;

exports[`verboseFormatter using cwd outputs 0 stdout column 1`] = `
"path/to/file.css
 1:2  ×  Unexpected foo  bar

1 problem (1 error, 0 warnings)

1 source checked
 /test-project/path/to/file.css

1 error found
 bar: 1"
`;

exports[`verboseFormatter using cwd outputs input CSS 1`] = `
"<input css>
 1:2  ×  Unexpected foo  bar

1 problem (1 error, 0 warnings)

1 source checked
 <input css>

1 error found
 bar: 1"
`;

exports[`verboseFormatter using cwd outputs less than 80 stdout column 1`] = `
"path/to/file.css
 1:2  ×  Unexpected foo  bar

1 problem (1 error, 0 warnings)

1 source checked
 /test-project/path/to/file.css

1 error found
 bar: 1"
`;

exports[`verboseFormatter using cwd outputs lineless syntax error 1`] = `
"path/to/file.css
      ×  Unexpected foo  SyntaxError

1 problem (1 error, 0 warnings)

1 source checked
 /test-project/path/to/file.css

1 error found
 SyntaxError: 1"
`;

exports[`verboseFormatter using cwd outputs no warnings 1`] = `
"1 source checked
 /test-project/path/to/file.css

0 problems found"
`;

exports[`verboseFormatter using cwd outputs one ignored file 1`] = `
"0 of 1 source checked
 file.css (ignored)

0 problems found"
`;

exports[`verboseFormatter using cwd outputs one warnings (of severity 'error') 1`] = `
"path/to/file.css
 1:2  ×  Unexpected foo  bar

1 problem (1 error, 0 warnings)

1 source checked
 /test-project/path/to/file.css

1 error found
 bar: 1"
`;

exports[`verboseFormatter using cwd outputs plugin rule warnings 1`] = `
"path/file.css
 1:2  ×  Unexpected foo  plugin/bar

1 problem (1 error, 0 warnings)

1 source checked
 /test-project/path/file.css

1 error found
 plugin/bar: 1"
`;

exports[`verboseFormatter using cwd outputs rule warnings with metadata 1`] = `
"path/file.css
 1:2  ×  Unexpected foo  no-foo

1 problem (1 error, 0 warnings)

1 source checked
 /test-project/path/file.css

1 error found
 no-foo: 1 (maybe fixable)

You may fix some problems with the \\"--fix\\" option."
`;

exports[`verboseFormatter using cwd outputs two of the same warnings of 'error' and one of 'warning' across two files 1`] = `
"path/to/file.css
 1:2  ×  Unexpected foo  bar
 2:3  ×  Unexpected foo  bar

path/to/file2.css
 3:1  ‼  Expected cat  baz

3 problems (2 errors, 1 warning)

2 sources checked
 /test-project/path/to/file.css
 /test-project/path/to/file2.css

2 errors found
 bar: 2

1 warning found
 baz: 1"
`;
